refactor febase to handle any type of basis

tarun-mitruka opened the pull request:
This PR is focused to handle the issue #241 to add handling of composite basis.
In this PR, `powerbasisfe.hh` and `scalarfe.hh` are merged to a file `febases.hh`. This is done to avoid duplication of code while adding another class to handle a composite basis. Due to this refactoring, the file `febases.hh` now resides outside - `ikarus/finiteelements/febases.hh` and the folder `ikarus/finiteelements/febases/autodifffe.hh` is now renamed to `ikarus/finiteelements/autodiff/autodifffe.hh`. `autodiffhelper.hh` will be moved from `utils` to this folder in a separate PR while handling #257.

Global indices can now handle, for example, the following basis types:

- `scalarBasis`
-  `powerBasis`
- `composite(scalarBasis, scalarBasis)` - For linear elastic Q1 elements, Timoshenko beam elements with higher order for $w$ compared to $\varphi$
- `composite(powerBasis, scalarBasis)` - For linear elastic Q1 with additional pressure DOFs
- `composite(scalarBasis, powerBasis)` - For Reissner-Mindlin plates with different order for $w$ compared to $\varphi_x, \varphi_y$
- `composite(powerBasis, powerBasis)` - For 7p-shells, advanced finite element technologies

A composite basis itself cannot have another composite basis though.
The existing finite elements itself will be refactored to accommodate the changes in this PR while handling the issue #244. 

!!! UPDATE 
Nested composite basis can also be handled after refactoring.
BlockedInterleaved can also be handled by the global indices.

tarun-mitruka committed:
refactor febases to handle compositeBasis

tarun-mitruka committed:
refactor to handle composite basis with more than two numChildren

tarun-mitruka committed:
rename variables

rath3t commented on ikarus/finiteelements/febases.hh:
```diff
@@ -0,0 +1,180 @@
+ * \tparam B The type of the basis.
+ */
+template <typename B>
+class FEBases
```
This class should handle in the end the flat global indices for all our finite elements. Thus, I think it should be better named: `FlatGlobalIndices`

This will be inline with my future refactoring of an fe skill system: https://godbolt.org/z/jjaEYMYY3

It will work like `makeBasis(gridView,preBasis)` such that we can write: `makeFE(linearElastic(5),skills(neumannBoundaryCondition(19),volumeLoad()))` 

then in the end this class we have here is a very fundamental "skill", which communicates the global indices: therfore i would propose the following name `FlatGlobalIndices`.

tarun-mitruka replied:
As we discussed, I moved `globalFlatIndices` function as a free function in `FEHelper` namespace. Then what do you think of the name `BasisSkills` for this class as it helps to provide all the underlying functionalities of a basis to a finite element?

tarun-mitruka replied:
I went with `FEBase` as the class name here as it is the base class from which all the finite elements must inherit from. 
I am not sure if it creates any confusion with the term `basis` here. 
The class itself gets the `globalBasis` and thereby binds the `localView` to the `element` and helps to provide some of the information related to the local view. May be we can also name the class as `LocalViewFE`? 
Since it denotes a base class to all the finite elements, may be we can also call it `FiniteElement`, from which all finite elements are derived, but may be then its purpose is not clear?
What do you think? @rath3t

rath3t commented on ikarus/finiteelements/febases.hh:
```diff
@@ -0,0 +1,180 @@
+   */
+  template <typename ChildIndex>
+  void globalScalarFlatIndices(std::vector<GlobalIndex>& globalIndices, const ChildIndex childIndex) const {
+    if constexpr (requires { localView_.tree().child(childIndex).finiteElement(); }) {
```
This code will never be used since you assume here anyway that your local view is scalar here.

tarun-mitruka replied:
I think this will be used if we have `compositeBasis(ScalarBasis)`, wouldn't it? Then there will be a child which is a scalar

rath3t replied:
Ah yes, maybe i didn't get the mechanism right.

rath3t commented on ikarus/finiteelements/febases.hh:
```diff
@@ -0,0 +1,180 @@
+  void globalCompositeFlatIndices(std::vector<GlobalIndex>& globalIndices, const ChildIndex childIndex) const {
+    constexpr int localNumChildren = LocalView::Tree::template Child<childIndex>::Type::degree();
+    if constexpr (localNumChildren == 0)
+      globalScalarFlatIndices(globalIndices, childIndex);
```
This should be merged into one.

rath3t commented on ikarus/finiteelements/febases.hh:
```diff
@@ -0,0 +1,180 @@
+    else if constexpr (isComposite) {
+      Dune::Hybrid::forEach(Dune::Hybrid::integralRange(Dune::index_constant<numChildren>()), [&](const auto i) {
+        static_assert(not(LocalView::Tree::template Child<i>::Type::isComposite),
+                      "Composite Basis cannot have a composite basis");
```
WE can call recursively again `globalCompositeFlatIndices(globalIndices, i);` here. Then it really works for all construction

rath3t replied:
We cann talk about that tomorrow.

rath3t commented on tests/src/testfebases.cpp:
```diff
@@ -0,0 +1,108 @@
+}
+
+template <typename TestSuiteType, typename... Bases>
+void checkBases(TestSuiteType& t, const std::vector<int>& expectedNumberOfChildren,
```
Can you create a new testSuite inside and return that one. Then inside you can also give it a better name.

rath3t commented on tests/src/testfebases.cpp:
```diff
@@ -0,0 +1,108 @@
+template <typename TestSuiteType, typename... Bases>
+void checkBases(TestSuiteType& t, const std::vector<int>& expectedNumberOfChildren,
+                const std::vector<int>& expectedSize, const std::string& basisName, const Bases&... bases) {
+  assert(sizeof...(bases) == expectedNumberOfChildren.size() && "Input size mismatch for number of children.");
```
Why do you you assert and not the testsuite?

rath3t commented on tests/src/testfebases.cpp:
```diff
@@ -0,0 +1,108 @@
+
+template <typename TestSuiteType, typename... Bases>
+void checkBases(TestSuiteType& t, const std::vector<int>& expectedNumberOfChildren,
+                const std::vector<int>& expectedSize, const std::string& basisName, const Bases&... bases) {
```
You know the size at compile time why use the runtime assert to checke the size of expectedSize. I would suggest creating a std::array

rath3t commented on tests/src/testfebases.cpp:
```diff
@@ -0,0 +1,108 @@
+  const auto power2 = getBasis(gridView, secondOrderLagrangePowerPreBasis);
+
+  /// Types of composite FE bases
+  const auto composite1 = getBasis(gridView, scalarScalarCompositePreBasis);
```
Lots of test cases here. Good!

 I would suggest we add a few nested ones. `power(composite(composite(...)), maybe from https://gitlab.dune-project.org/staging/dune-functions/-/blob/master/dune/functions/functionspacebases/test/makebasistest.cc?ref_type=heads and fix the code. Then we never have to touc hit again.

Additionall we should add one test with vector-valued fe such as nedelc. To really have the correct forever.

rath3t commented on tests/src/testfebases.cpp:
```diff
@@ -0,0 +1,108 @@
+        checkScalars(t, static_cast<int>(bases.size()), expectedSize[i],
+                     " Size is incorrect for i = " + std::to_string(i) + basisName);
+        std::vector<typename std::remove_cvref_t<decltype(bases.localView())>::MultiIndex> dofs;
+        bases.globalFlatIndices(dofs);
```
You can check a lot more here. 
First you can sort the indices with `std::ranges::sort` and then check for duplicates with 
```cpp
bool hasDuplicates = std::adjacent_find(sortedVector.begin(), sortedVector.end()) != sortedVector.end();
```

rath3t reviewed (commented):
Lets talk about the generalization of this code. Thanks!

tarun-mitruka committed:
refactoring as per PR review

tarun-mitruka committed:
add forEachLeafOrPowerNode

tarun-mitruka committed:
refactoring and fixes

tarun-mitruka committed:
fix doxygen comment

rath3t commented on ikarus/finiteelements/febases.hh:
```diff
@@ -0,0 +1,77 @@
+   *
+   * \return The number of children of the basis.
+   */
+  [[nodiscard]] int numberOfChildren() const { return localView_.tree().degree(); }
```
This function can be removed. If Someone is interested it can be derived from the localView and simplifies this interface.

rath3t commented on ikarus/finiteelements/fehelper.hh:
```diff
@@ -38,4 +38,80 @@ auto localSolutionBlockVector(const typename Traits::FERequirementType::Solution
+  if constexpr (Tree::isLeaf) {
+    leafFunc(tree);
+  } else if constexpr (Tree::isPower) {
+    if constexpr (Tree::template Child<Dune::Indices::_0>::Type::isComposite) {
```
This is a strange statement it should be the other way around. Check if the child of the powernode is leaf. If not call forEachLeafOrPowerNode again with the child.

tarun-mitruka replied:
If I check if the child of a power node is a leaf and call the `leafFunc`, then it will no longer be in `FlatInterleaved` format right? Because then each leaf node of a power node will be looped over bringing in `x0,x1,...,xn,y0,y1,...,yn` format? In order to invoke `powerFunc` at a power node directly and have `FlatInterleaved` format, I chose the other way around of checking if it is not a leaf. May be we can discuss this together.

rath3t replied:
Not you should check if the child of the power node is leaf and then pass the power node.

rath3t replied:
Sinece we only want to call the `powerNodeOp` for power nodes, which has leaf children. So only for `power(lagrange,d)` and  `power(nedelec,d)`  but not for `power(composite(nedelec, power(lagrange,d2)),d)`

rath3t replied:
Then if you pass the power node int o`powerNodeOp` you only implement the `if` branch. We can also rename the function name to `forEachLeafOrPowerLeafNode`

rath3t commented on ikarus/finiteelements/fehelper.hh:
```diff
@@ -38,4 +38,80 @@ auto localSolutionBlockVector(const typename Traits::FERequirementType::Solution
+ * \param leafFunc A functor to be called for a leaf node.
+ */
+template <class T, class ChildIndex, class PowerFunc, class LeafFunc>
+void forEachLeafOrPowerNode(T&& tree, ChildIndex&& childIndex, PowerFunc&& powerFunc, LeafFunc&& leafFunc) {
```
Why are you using this childIndex? This is not needed I suppose. Simply pass the child. Additionally the signature would be better if you just pass the `treePath` as in https://gitlab.dune-project.org/staging/dune-typetree/-/blob/master/dune/typetree/traversal.hh?ref_type=heads#L164. Why did you modify the signature?

rath3t commented on ikarus/finiteelements/fehelper.hh:
```diff
@@ -38,4 +38,80 @@ auto localSolutionBlockVector(const typename Traits::FERequirementType::Solution
+ * \param globalIndices Output vector to store global indices.
+ */
+template <typename LocalView>
+void globalFlatIndices(const LocalView& localView, std::vector<typename LocalView::MultiIndex>& globalIndices) {
```
I think it is better to make this function simply receive the finite Element. Then a user can provide an overload of this function.

rath3t replied:
And the namespace name FEHelper is also very generic. Maybe without this additional namespace?

tarun-mitruka replied:
Passing the finite element is may be better I guess and then the local view can be accessed simply via `fe.localView()`.
Since it helps the finite elements to manage its global indices, may be the namespace is okay, but I do the following from the other comment?
```cpp
using FEHelper::globalFlatIndices;
globalFlatIndices(fe.localView(), dofs);
```

rath3t replied:
Not sure what your question is actually at the end here. We can talk about it later.

rath3t commented on ikarus/assembler/simpleassemblers.inl:
```diff
@@ -164,7 +164,7 @@ void SparseFlatAssembler<B, FEC>::createOccupationPattern(Eigen::SparseMatrix<do
   for (auto&& fe : this->finiteElements()) {
     dofs.resize(0);
-    fe.globalFlatIndices(dofs);
+    FEHelper::globalFlatIndices(fe.localView(), dofs);
```
If we have a namespace it would better to use 

```cpp
using FEHelper::globalFlatIndices;
globalFlatIndices(fe.localView(), dofs);
```

to enable ADL for other functions.

rath3t commented on ikarus/finiteelements/fehelper.hh:
```diff
@@ -38,4 +38,80 @@ auto localSolutionBlockVector(const typename Traits::FERequirementType::Solution
+
+  /** \brief Functor to obtain global indices of a power node. */
+  auto powerOpFunc = [&](auto&& node, auto&& ci) {
+    if constexpr (requires { node.child(ci).finiteElement(); }) {
```
Can you elaborate again what both of these branches taking care of? The else branch is the usual `power(lagrange,d)` but for which case is the first used?

tarun-mitruka replied:
The `if` branch is used if we have `power(lagrange, d)` for example. The `else` branch is used when we have, for example, `composite(lagrange, power(lagrange))`. Then the finite element can be accessed via `child(Dune::Indices::_1, 0).finiteElement()`.

tarun-mitruka replied:
To have this differentiation, I have used `childIndex` in the function `forEachLeafOrPowerNode`. This made having `treePath` redundant there and I have removed it. I have to check if it is possible to access the child correctly with the tree path.

rath3t replied:
Ok thanks for the explaination. But the powerOpFunc should receive only the powerNode and not the rootNode. This complicates everything. You should directly pass the child into this function. Then this only uses the `if` branch and the other boilerplate can be removed, which is unnecessary code duplciation.

rath3t commented on tests/src/testfebases.cpp:
```diff
@@ -0,0 +1,164 @@
+                 " Number of children is incorrect" + message);
+    checkScalars(t, static_cast<int>(basis.size()), expectedSize[i], " Size is incorrect" + message);
+    std::vector<MultiIndex> dofs;
+    Ikarus::FEHelper::globalFlatIndices(basis.localView(), dofs);
```
Actually `globalFlatIndices` would exactly look the same for blocked basis. We should rename the function to globalIndices. Can you also provide examples fro the blocked case? Yo simply have to make the getBasis function above with untouched instead of flat.

rath3t commented on tests/src/testfebases.cpp:
```diff
@@ -0,0 +1,164 @@
+template <typename GridView, typename PreBasis>
+auto getBasis(const GridView& gridView, const PreBasis& preBasis) {
+  const auto basis      = Ikarus::makeBasis(gridView, preBasis);
+  const auto& localView = basis.flat().localView();
```
You can use an imediate evaluated lambda here, because otherwise you can not do this inside a if constexpr branch.

```cpp
const auto& localView = [&]() -> const  auto& {
if constexpr (useFlat)
   return basis.flat().localView();
else
   return basis.untouched().localView();
}();

rath3t replied:
And all the example bases shoujld run with the non-type template `bool useFlat=false/true`

rath3t reviewed (commented):
This is evolving nicely but there are still things to make easier. We can also look at it together. Furthermore, we also have to modify `src/iks003_incompressibleLinearElasticity.cpp`because this breaks now, doesnt it? Maybe we want to make the function there a free function for the educational aspect or remove it completly.

It actually doesnt break but the `globalFlatIndices` member function is stale now and unused. Maybe this also has to be  refactored in the docs?

rath3t commented on ikarus/finiteelements/febases.hh:
```diff
@@ -0,0 +1,77 @@
+namespace Ikarus {
+
+/**
+ * \brief FEBase class for working with different types of bases.
```
Outdated comment.

rath3t commented on ikarus/finiteelements/febases.hh:
```diff
@@ -0,0 +1,77 @@
+
+/**
+ * \file febases.hh
+ * \brief Contains the FEBase class, which works with different types of bases.
```
Outdated comment.

rath3t commented on docs/website/02_examples/vonMisesTruss.md:
````diff
@@ -7,13 +7,13 @@ standard Von-Mises truss example found in literature (refer to Section 2[@misesT
 ```cpp
 Truss(const Basis &basis, const typename LocalView::Element &element, double p_EA)
-    : BaseDisp(basis, element), EA{p_EA} {
+    : FEBase<Basis, useFlat>(basis, element), EA{p_EA} {
````
Maybe FEBase should have true as default and then here only FEBase<Basis> will be left. Otherwise this random name `useFlat`  is confusing here.

tarun-mitruka replied:
I thought in the future, may be it is nicer to have the flexibility of using either the flat or the untouched version of the basis from the finite element side, and hence I am passing it explicitly from the finite element and did not use the default provided in `FEBase`. Else, the non-type template paramater `useFlat` is not needed for the finite elements at all and it will always assume `FlatBasis`. 
In a separate PR, may be then `using FlatBasis = typename Traits::FlatBasis;` can also be replaced to `using BasisVersion= typename Traits::BasisVersion;`
What do you think?

rath3t replied:
Yes you are right but I'm concert that this makes our documentation confusing. Maybe you can refactor such that:
E.g. 
```cpp
template<tyname Basis, bool useFlat>
class Truss: FEBase<Basis, useFlat>
{
using Base = FEBase<Basis, useFlat>;

Truss(const Basis &basis, const typename LocalView::Element &element, double p_EA)
    : Base(basis, element), EA{p_EA} {
...
}
```

Then the constructor looks nicer

tarun-mitruka replied:
Ah okay. I will then fix it in `ikarus-examples` and amend the corresponding documentation. I think the finite elements in `ikarus` can stay as such.

rath3t replied:
Ok for me

rath3t commented on ikarus/finiteelements/mechanics/kirchhoffloveshell.hh:
```diff
@@ -34,22 +34,23 @@ namespace Ikarus {
 {
 public:
-  using Traits            = FETraits<B, FER, useEigenRef>;
+  using Traits            = FETraits<B, useFlat, FER, useEigenRef>;
```
This simply should be `using Traits = FEBase<B, useFlat>::Traits;`

tarun-mitruka replied:
If we do it like this, `useEigenRef` is not used at all?

rath3t replied:
Ah i made a small mistake but yes that the point all the arguments should be forwarded to `FEBase<B, useFlat,FER,useEigenRef>; and all the other elements just use the traits from the base class. Therefore, all the arguments should be passed, I don't know why you left the arguments as default.

rath3t commented on ikarus/finiteelements/mechanics/nonlinearelastic.hh:
```diff
@@ -40,22 +40,23 @@ namespace Ikarus {
 {
 public:
-  using Traits            = FETraits<B, FER, useEigenRef>;
+  using Traits            = FETraits<B, useFlat, FER, useEigenRef>;
```
also here

rath3t commented on ikarus/finiteelements/mechanics/linearelastic.hh:
```diff
@@ -41,22 +41,23 @@ namespace Ikarus {
 {
 public:
-  using Traits               = FETraits<B, FER, useEigenRef>;
+  using Traits               = FETraits<B, useFlat, FER, useEigenRef>;
```
here

rath3t commented on ikarus/utils/traversal.hh:
```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+#pragma once
+
```
file doxygen header missing here

rath3t commented on ikarus/utils/traversal.hh:
```diff
@@ -0,0 +1,46 @@
+
+/**
+ * \brief A function which loops over all the nodes of a tree and performs different actions
+ * for a power node and a leaf node depending on the corresponding functor passed.
```
Can you also indicate that this is not for all power nodes but only for those with leafs as childs.

rath3t commented on ikarus/finiteelements/fetraits.hh:
```diff
@@ -32,11 +33,17 @@ struct FETraits
+  using UntouchedBasis = typename Basis::UntouchedBasis;
+
+  /** \brief Type of the basis version*/
+  using BasisVersion = std::conditional_t<useFlat, FlatBasis, UntouchedBasis>;
```
Maybe stick to `Basis` as name, since most of the code should care about what type and this is the best one

tarun-mitruka replied:
`Basis` already existed there which is the type of the global basis passed to a finite element. Hence I used BasisVersion here. Renaming it to `Basis` might then create confusions in finite elements. What about just `PreBasis`?

rath3t replied:
It is no PreBasis, maybe we have to rename the Ikarus::Basis to Ikarus::BasisHandler to not confuse it with the Dune::Basis stuff?

tarun-mitruka replied:
You mean here `ikarus/utils/basis.hh` right? Makes sense, then `FETraits` will get `GB` instead of `B`, where `GB` is then the `GlobalBasis` which then helps to rename `BasisVersion` to just `Basis`

rath3t replied:
FETraits will get `BH ` for basis handler? Inside you extract then untouched and flat

>You mean here ikarus/utils/basis.hh right?

Yes

tarun-mitruka replied:
I am not renaming the `makeBasis` function to `makeBasisHandler` here, may be it can be done separately. I have renamed the file to `basishandler.hh`

tarun-mitruka replied:
Since this modification touches all the tests and examples, I will do it in a separate PR, right after this PR is merged via #268

rath3t reviewed (changes requested):
We are almost done now. Thanks for your patience!

rath3t commented on ikarus/finiteelements/mechanics/kirchhoffloveshell.hh:
```diff
@@ -34,22 +34,23 @@ namespace Ikarus {
-                           public Traction<KirchhoffLoveShell<B, FER, useEigenRef>, FETraits<B, FER, useEigenRef>>
+template <typename B, typename FER = FERequirements<>, bool useEigenRef = false, bool useFlat = true>
+class KirchhoffLoveShell
+    : public FEBase<B, useFlat>,
```
E.g. here all args should be forwarded otherwise everything will break.

rath3t commented on docs/website/02_examples/incompressibleRubberBlock.md:
```diff
@@ -8,21 +8,22 @@ for such a system is defined in the `Solid struct` by the function
-  lambdaMat = convertLameConstants({.emodul = emod_, .nu = nu_}).toLamesFirstParameter();
+Solid(const BasisHandler &basis, const typename LocalView::Element &element, double emod, double nu)
+      : Base(basis, element), emod_{emod}, nu_{nu} {
+    this->localView().bind(element);
```
Actually This bind command is not needed since the bind happens in the base class anyway.

rath3t replied:
Can also removed in ther KL plate code

rath3t commented on docs/website/02_examples/vonMisesTruss.md:
````diff
@@ -7,13 +7,13 @@ standard Von-Mises truss example found in literature (refer to Section 2[@misesT
 ```cpp
-Truss(const Basis &basis, const typename LocalView::Element &element, double p_EA)
-    : BaseDisp(basis, element), EA{p_EA} {
+Truss(const BasisHandler &basis, const typename LocalView::Element &element, double p_EA)
````
The argument name should also be not `basis` anymore

rath3t commented on ikarus/finiteelements/febase.hh:
```diff
@@ -0,0 +1,80 @@
+   * \param basis The basis.
+   * \param element The local element.
+   */
+  explicit FEBase(const BH& basis, const typename LocalView::Element& element)
```
`basis` -->

rath3t commented on ikarus/finiteelements/mechanics/kirchhoffloveshell.hh:
```diff
@@ -99,11 +102,11 @@ public:
+      : Base(basisHandler, element),
         VolumeType(volumeLoad),
         TractionType(neumannBoundary, neumannBoundaryLoad),
         emod_{emod},
```
Below the bind can be removed

rath3t commented on ikarus/finiteelements/mechanics/linearelastic.hh:
```diff
@@ -76,10 +78,10 @@ public:
+      : Base(basisHandler, element),
         VolumeType(volumeLoad),
         TractionType(neumannBoundary, neumannBoundaryLoad),
         emod_{emod},
```
here too

rath3t commented on ikarus/finiteelements/mechanics/nonlinearelastic.hh:
```diff
@@ -69,18 +71,18 @@ public:
                    VolumeLoad volumeLoad = {}, const BoundaryPatch<GridView>* neumannBoundary = nullptr,
                    NeumannBoundaryLoad neumannBoundaryLoad = {})
-      : BasePowerFE(globalBasis, element),
+      : Base(basisHandler, element),
```
also here

rath3t reviewed (changes requested):
Small changes. Please change the argument name according to the refactor and remove unnecessary localview binds

rath3t reviewed (approved):
Ok looks good if the ci runs through

Commit:
refactoring as per PR comments

new fixes as per PR comments

final fixes

fix docs according to fix in codestyle in ikarus-examples
